clvm-derive/src/from_clvm.rs

```diff
@@ -1,66 +1,313 @@
 use proc_macro2::{Ident, Span, TokenStream};
 use quote::quote;
 use syn::{
-    parse_quote, spanned::Spanned, Data, DeriveInput, Fields, GenericParam, Type, TypeParam,
+    parse_quote, spanned::Spanned, Data, DeriveInput, Expr, Fields, FieldsNamed, FieldsUnnamed,
+    GenericParam, Type,
 };
 
-use crate::helpers::{add_trait_bounds, parse_args, Repr};
+use crate::{
+    helpers::{add_trait_bounds, parse_clvm_attr, parse_int_repr, Repr},
+    macros::{repr_macros, Macros},
+};
 
-pub fn from_clvm(mut ast: DeriveInput) -> TokenStream {
-    let args = parse_args(&ast.attrs);
-    let crate_name = quote!(clvm_traits);
+#[derive(Default)]
+struct FieldInfo {
+    field_types: Vec<Type>,
+    field_names: Vec<Ident>,
+    initializer: TokenStream,
+}
+
+struct VariantInfo {
+    name: Ident,
+    value: Expr,
+    field_info: FieldInfo,
+    macros: Macros,
+}
 
-    let field_types: Vec<Type>;
-    let field_names: Vec<Ident>;
-    let initializer: TokenStream;
+pub fn from_clvm(ast: DeriveInput) -> TokenStream {
+    let clvm_attr = parse_clvm_attr(&ast.attrs);
+    let crate_name = quote!(clvm_traits);
 
     match &ast.data {
-        Data::Struct(data_struct) => match &data_struct.fields {
-            Fields::Named(fields) => {
-                let fields = &fields.named;
-                field_types = fields.iter().map(|field| field.ty.clone()).collect();
-                field_names = fields
-                    .iter()
-                    .map(|field| field.ident.clone().unwrap())
-                    .collect();
-                initializer = quote!(Self { #( #field_names, )* });
+        Data::Struct(data_struct) => {
+            if clvm_attr.untagged {
+                panic!("cannot use `untagged` on a struct");
```

Comment:
I think it would be great to have tests for these compilation failures. I believe it could be done using [compiletest-rs](https://github.com/Manishearth/compiletest-rs)